include/kernels/copy_field_offset.cuh

```diff
@@ -0,0 +1,251 @@
+      }
+
+      for (int d = 0; d < nDim; d++) {
+        dim_out[d] = out_field.full_dim(d);
+        dim_in[d] = in_field.full_dim(d);
+      }
+
+      X0h_out = dim_out[0] / 2;
+      X0h_in = dim_in[0] / 2;
+
+      volume_cb_in = in_field.VolumeCB();
+      volume_cb_out = out_field.VolumeCB();
+
+      if (volume_cb_out > volume_cb_in) {
+        volume_cb = volume_cb_in;
+        mode = QudaOffsetCopyMode::COLLECT;
+      } else {
+        volume_cb = volume_cb_out;
+        mode = QudaOffsetCopyMode::DISPERSE;
+      }
+
+      volume_4d_cb_in = volume_cb_in / Ls;
+      volume_4d_cb_out = volume_cb_out / Ls;
+      volume_4d_cb = volume_cb / Ls;
+    }
+  };
+
+  template <class Arg>
+  __device__ __host__ inline
+    typename std::enable_if<std::is_same<typename Arg::Field, ColorSpinorField>::value, void>::type
+    copy_field(int out, int in, int parity, Arg &arg)
+  {
+    using Element = typename Arg::Element;
+
+    Element element = arg.in(in, parity);
+    arg.out(out, parity) = element;
+  }
+
+  template <class Arg>
+  __device__ __host__ inline typename std::enable_if<std::is_same<typename Arg::Field, CloverField>::value, void>::type
+  copy_field(int out, int in, int parity, Arg &arg)
+  {
+    using Element = typename Arg::Element;
+    constexpr int length = 72;
+    Element reg_in[length];
+    Element reg_out[length];
+    arg.in.load(reg_in, in, parity);
+    for (int i = 0; i < length; i++) { reg_out[i] = reg_in[i]; }
+    arg.out.save(reg_out, out, parity);
+  }
+
+  template <class Arg>
+  __device__ __host__ inline typename std::enable_if<std::is_same<typename Arg::Field, GaugeField>::value, void>::type
+  copy_field(int out, int in, int parity, Arg &arg)
+  {
+    using Element = typename Arg::Element;
+#pragma unroll
+    for (int d = 0; d < 4; d++) {
+      Element element = arg.in(d, in, parity);
+      arg.out(d, out, parity) = element;
+    }
+  }
+
+  template <QudaOffsetCopyMode mode, class Arg>
+  __device__ __host__ void copy_field_offset(int x_cb, int s, int parity, Arg &arg)
+  {
+    // XXX: This code assumes parity if NOT changed when offset is added.
+    if (Arg::pc_type == QUDA_4D_PC) { // 4d even-odd preconditioning, works for most fermions
+      int coordinate[4];
+      int idx_in;
+      int idx_out;
+      if (mode == QudaOffsetCopyMode::COLLECT) {
+        // we are collecting so x_cb is the index for the input.
+        idx_in = x_cb;
+        getCoordsExtended(coordinate, x_cb, arg.dim_in, parity, arg.offset.array);
+        idx_out = linkIndex(coordinate, arg.dim_out);
+      } else {
+        // we are dispersing so x_cb is the index for the output.
+        idx_out = x_cb;
+        getCoordsExtended(coordinate, x_cb, arg.dim_out, parity, arg.offset.array);
+        idx_in = linkIndex(coordinate, arg.dim_in);
+      }
+      copy_field(s * arg.volume_4d_cb_out + idx_out, s * arg.volume_4d_cb_in + idx_in, parity, arg);
+    } else { // 5d even-odd preconditioning, works for 5d DWF
```

Comment:
Is there something specific we can put for an `else if` here, and then have a `static_assert` at the top of this function making sure `Arg::pc_type` matches all the possibilities? (As someone who's learned the hard way what happens when you assume an `else` catches all your problems ;) )